median absolute deviation agg

andyb-elastic opened the pull request:
This commit adds a new single value metric aggregation that calculates
the statistic called median absolute deviation, which is a measure of
variability that works on more types of data than standard deviation

Our calculation of MAD is approximated using t-digests. In the collect
phase, we collect each value visited into a t-digest. In the reduce
phase, we merge all value t-digests, then create a t-digest of
deviations using the first t-digest's median and centroids

For #26681

Still writing the docs, but wanted to get feedback on the code

elasticmachine commented:
Pinging @elastic/es-search-aggs

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,117 @@
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalMedianAbsoluteDeviation magAgg = (InternalMedianAbsoluteDeviation) aggregation;
+            if (valueMerged == null) {
+                valueMerged = new TDigestState(magAgg.valuesSketch.compression());
```
To avoid this check I think we can just initialise the `valueMerged` to an empty T-Digest outside the loop and merge each response into that?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregator.java:
```diff
@@ -0,0 +1,157 @@
+        Releasables.close(valueSketches);
+    }
+
+    public static double computeMedianAbsoluteDeviation(TDigestState valuesSketch) {
```
Personally I think this static method should live in `InternalMedianAbsoluteDeviation`.

polyfractal replied:
++ Seems more like an `Internal`'y method, even though it's used in multiple places

andyb-elastic replied:
👍

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,117 @@
+
+    @Override
+    public double getMedianAbsoluteDeviation() {
+        return computeMedianAbsoluteDeviation(valuesSketch);
```
This is a relatively expensive operation, I think we should calculate this value once and store it rather than calculating it each time the getter is called, especially as the value cannot change. Maybe we should calculate this value in the constructor and store it along side the sketch or calculate it in the reduce method and store it alongside the sketch at that point?

andyb-elastic replied:
Definitely agree, I had a couple ways of doing it before but nixed them because they were gross. Don't know why doing it in the constructor didn't occur to me, that's much cleaner

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,117 @@
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        TDigestState valueMerged = null;
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalMedianAbsoluteDeviation magAgg = (InternalMedianAbsoluteDeviation) aggregation;
```
From the code in The aggregator factory below it looks like `magAgg.valuesSketch` can be null if the agg is unmapped, I think we need to guard against NPEs here?

andyb-elastic replied:
I'm not sure, I think it gets protected against the valuesSketch being null in [`MedianAbsoluteDeviationAggregator#buildAggregation`](https://github.com/elastic/elasticsearch/pull/34482/files#diff-a92fa3984e0cd1c5e4f9688b5b6cf914R121) because if we never built a sketch for the bucket it's requesting, we just create an empty one. Is there another way the internal aggregation gets constructed?

andyb-elastic replied:
The typo definitely needs fixing though

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregator.java:
```diff
@@ -0,0 +1,157 @@
+
+    public static double computeMedianAbsoluteDeviation(TDigestState valuesSketch) {
+
+        if (valuesSketch.size() == 0) {
```
Since unmapped aggs have a `null` sketch I think we need a null check in this method?

colings86 replied:
I assume the unmapped and partially unmapped tests in the integration test fail at the moment because of this? If not, we should work out why and ensure they are correct and have a test that tests for NPEs here

andyb-elastic replied:
This method actually has the same behavior (returning NaN) without this check, because we create an empty `approximatedDeviationsSketch`, and calling `#quantile` on an empty sketch returns NaN. I added this check to make it really obvious what happens without data points, since I don't think that `#quantile` returning NaN is documented

andyb-elastic replied:
I explained why I don't think `valuesSketch` can be null here in another comment, the unmapped and partially unmapped tests were passing when I committed this

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregationBuilder.java:
```diff
@@ -0,0 +1,140 @@
+     * Set the compression factor of the t-digest sketches used
+     */
+    public MedianAbsoluteDeviationAggregationBuilder compression(double compression) {
+        if (compression < 0.0) {
```
Not really a MAD issue just something that caught my eye.  I know we allow zero as a valid compression in TDigest percentiles... but I'm wondering if we shouldn't allow it?

A quick skim of AVLTreeDigest shows `compression` used as a [denominator](https://github.com/tdunning/t-digest/blob/master/core/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L121) and in a [multiplication](https://github.com/tdunning/t-digest/blob/master/core/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L154).  The agg still "works" with `0` compression, but I'm thinking that will cause a compression on each new point?  Should we allow that sort of behavior?

colings86 replied:
+1 to not allowing compression of 0, it seems dangerous and trappy to me

andyb-elastic replied:
Definitely agree

andyb-elastic commented:
I made some changes to make more clear what is expected to be null or non null in the internal agg + aggregator

andyb-elastic commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(valuesSketch, medianAbsoluteDeviation);
```
Not sure if the MAD should be included here and equals 🤔 

I believe equal values sketches implies equal MAD

colings86 replied:
Agreed

andyb-elastic commented:
Also looks like we have some [yaml rest tests for percentiles](https://github.com/elastic/elasticsearch/blob/48ab6d7c92180285f5777ebd157463163c5600f9/rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/180_percentiles_tdigest_metric.yml), would it be appropriate to add some for this agg?

andyb-elastic commented:
Jenkins run gradle build tests please

colings86 commented:
> Also looks like we have some yaml rest tests for percentiles, would it be appropriate to add some for this agg?

Yes I think we should have YAML tests for this aggregation to test that the aggregation works end to end as expected

andyb-elastic commented:
I added a yaml test

Another change I pushed worth noting - before I opened this I changed all of the class names to use the full "median absolute deviation" rather than the abbreviation "mad". Today I realized I'd forgotten to do it with the dsl name for the agg too, so it's now `median_absolute_deviation`. I like that for consistency but it is a little more verbose than the other aggs

andyb-elastic commented:
It looks like the failures for aeaf714 in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/707/ have been occurring in master so I think they're unrelated

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,137 @@
+    static double computeMedianAbsoluteDeviation(TDigestState valuesSketch) {
+
+        if (valuesSketch.size() == 0) {
+
```
Crazy small nit: extra spaces :)

andyb-elastic replied:
Ha, this spacing (and the other place you pointed it out) is actually intentional because it seems more readable to me. But I agree that it's not consistent with the dominant style, will fix

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalMedianAbsoluteDeviation.java:
```diff
@@ -0,0 +1,137 @@
+    protected final TDigestState valuesSketch;
+    protected final double medianAbsoluteDeviation;
+
+    public InternalMedianAbsoluteDeviation(String name,
```
I think this can be package-private, and the two class variables (sketch and deviation) private?

andyb-elastic replied:
👍

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregationBuilder.java:
```diff
@@ -0,0 +1,140 @@
+    public MedianAbsoluteDeviationAggregationBuilder compression(double compression) {
+        if (compression <= 0d) {
+            throw new IllegalArgumentException(
+                "[compression] must be greater than 0. Found [" + compression + "] in [" + name + "]");
```
Tiny nit: could we use the ParseField here?  That way if we ever change the name we won't forge to also change the exception

andyb-elastic replied:
Good call

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregator.java:
```diff
@@ -0,0 +1,147 @@
+        this.format = Objects.requireNonNull(format);
+
+        this.compression = compression;
+
```
Please don't hate me nit: whitespace around these assignments :)

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregator.java:
```diff
@@ -0,0 +1,147 @@
+
+    private ObjectArray<TDigestState> valueSketches;
+
+    public MedianAbsoluteDeviationAggregator(String name,
```
I think this can be package-private

andyb-elastic replied:
👍

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregatorFactory.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.List;
+import java.util.Map;
+
+public class MedianAbsoluteDeviationAggregatorFactory extends
```
I think this can be package-private too

andyb-elastic replied:
👍

polyfractal commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationIT.java:
```diff
@@ -0,0 +1,621 @@
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsNull.notNullValue;
+
+public class MedianAbsoluteDeviationIT extends AbstractNumericTestCase {
```
I love that we have these tests, but I wonder if we should move them to yml tests?  Or maybe into the `AggregatorTestCase` tests?  Just because we've been discouraging IT tests unless necessary due to their slowness in the build.  I think we'd lose the script tests though, since the yml runner doesn't include script module iirc.

Unless the desire was to index a bunch of docs to test accuracy, in which case maybe it should be a QA test?

@colings86 wdyt?

colings86 replied:
I think we should keep this for now since its the only way to test the script capabilities of the aggregation and that's an important piece of functionality for all aggregations to have.

polyfractal replied:
Right, makes sense.  :+1:

andyb-elastic replied:
Scripting aside, I'm not totally sold on the method for testing accuracy here (and in the aggregator test case) - I just wanted something a little more precise than "is this even a reasonable answer". If this ends up failing in CI for not meeting the accuracy cutoff, we can reduce it or test accuracy in a different way

polyfractal commented on rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/270_median_absolute_deviation_metric.yml:
```diff
@@ -0,0 +1,122 @@
+setup:
```
Could we add a test where there are no docs (or no fields) that match, so we can verify the xcontent spits out a `null` value when nothing was collected?

andyb-elastic replied:
Yeah this is definitely important, thanks for catching

andyb-elastic commented:
@polyfractal thanks for the comments, I think I've addressed all of them. I'll ping again after pushing the docs

polyfractal commented on docs/reference/aggregations/metrics/median-absolute-deviation-aggregation.asciidoc:
```diff
@@ -0,0 +1,185 @@
+is calculated as the median of each data point's deviation from the sample
+median. That is, for a random variable X, the median absolute deviation is
+median(|median(X) - X~i~|).
+
```
Perhaps a sentence (or addition to existing sentence) somewhere that describes MAD as being robust/resistant to change by outliers?

andyb-elastic replied:
👍

polyfractal commented on docs/reference/aggregations/metrics/median-absolute-deviation-aggregation.asciidoc:
```diff
@@ -0,0 +1,185 @@
+treated. By default they will be ignored but it is also possible to treat them
+as if they had a value.
+
+Let's be optimistic and assume some reviewers loved the product so much that
```
:D

polyfractal commented:
> Discussion of the approximation accuracy resulting from the benchmarking I did - while I'd like to give the user a sense of how it performs on the distributions we tested it seems irresponsible to make claims about it given that we won't be retesting it over time

Just a comment on this; I think it's fine to give some indication of how approximate the algo is, just so the user has some notion of what to expect.  Even if we don't continually run the benchmarks it might be useful for them to know the ballpark.  "1-5%", etc is about as specific as we get for the other approximate aggs (and disclaimers about being dependent on data)

andyb-elastic commented:
Thanks, I wrote essentially that. In my benchmarking we always got within 1% with compression = 1000 (what I set the default as) but I noted 5% to be safe

andyb-elastic commented:
Jenkins run gradle build tests please

andyb-elastic commented:
master: https://github.com/elastic/elasticsearch/commit/b8280ea7cceb5f31c2827f09a49b18eaeb9ddebd
6.x: https://github.com/elastic/elasticsearch/commit/3fd9dd4f09bfde3194c710b2047cf7c3c6d5baa4
